### PR TITLE
hmem: remove cuda-specific functions when using `ofi_hmem_dev_reg_*`

### PR DESCRIPTION
**Pull request by thomasgillis**
remove cuda-specific naming and functions when using `ofi_hmem_dev_reg_*` function family
- [x] removes the need to check for `cuda_is_gdrcopy_enabled()`: `ofi_hmem_dev_register` fails if GDR copy is not available
- [x] update the bit mask to remove reference to cuda
- [x] ~update the EFA provider accordingly (@wenduwan do you want to do it here or in another PR?)~

### CODE REVIEW

**Comment by shijin-aws** [comment]
AFAIK neuron doesn't support this yet. It is safe to make it as ENOSYS

**Comment by thomasgillis** [comment]
what should I use for `rocm` and `ze`?

**Comment by shijin-aws** [comment]
> Adds a new operation to `dev_reg` path:
> 
>     * [x]  system
> 
>     * [x]  cuda
> 
>     * [ ]  rocr
> 
>     * [ ]  ze
> 
>     * [ ]  others

AFAIK neuron doesn't support this yet. It is safe to make it as false. Likely the same for synapseai

**Comment by shijin-aws** [comment]
>what should I use for rocm and ze?

I would check with @j-xiong  for ze, and @edgargabriel  for rocm

**Comment by j-xiong** [comment]
I think it's fine to set ze and rocm to the default "no" variation. Both ze and rocm has the dev_reg function added, but we may want some control over whether or not to enable gdrcopy. We can later enbale ze and rocm seperately.

**Comment by thomasgillis** [comment]
@edgargabriel the current version of `rocr_dev_register` simply gets the device and host pointer using
https://github.com/ofiwg/libfabric/blob/14594abdcf3b664eece27e4978efd1ae82497a94/src/hmem_rocr.c#L995-L996

is there any good side effect in enabling it by default?

**Comment by thomasgillis** [comment]
Ok sounds good, will wait for his input then.
Should I introduce an environment variable? Or a default to always `true`?

File: include/ofi_hmem.h [hunk 1]
```diff
@@ -120,6 +120,7 @@ struct ofi_hmem_ops {
 			     size_t *base_length);
 	bool (*is_ipc_enabled)(void);
 	int (*get_ipc_handle_size)(size_t *size);
+	bool (*is_dev_reg_enabled)(void);
```
Comment (iziemba) [inline include/ofi_hmem.h:123]: Hi Thomas,

Could you explain the use-case for this and why the current dev_reg/unreg APIs will not work?

The use-case that dev_reg/unreg enables is for providers to cache a dev_reg handle with their memory registration. During memory registration, if dev_reg fails, the expectation is the provider implements a fallback. I would expect the fallback to be using the non-dev_reg memcpy APIs which could hurt performance.

Comment (thomasgillis) [reply include/ofi_hmem.h:123]: sure, so all this came from the need of adding `cuda gdrcopy` to `rxm` (as done already in `efa`) when doing injection/inlining for send/recv. With cuda, `gdrcopy` is used through the `dev_reg` features and is conditionally registered at mr time (with cuda_is_gdrcopy_enabled):

https://github.com/ofiwg/libfabric/blob/c37c44e5b485b5185de96c53453a7249ca3fd672/prov/rxm/src/rxm_domain.c#L602-L613

https://github.com/ofiwg/libfabric/blob/9f604d658277c3a4851506327cb7a75572622767/prov/efa/src/efa_mr.c#L253-L256

one of the comment in [this PR](https://github.com/ofiwg/libfabric/pull/9454#discussion_r1369355045) highlighted that it doesn't make sense to keep the conditional in cuda as the `dev_reg` feature should be used by all the iface (rocr, cuda, ze, etc), hence the present draft PR.
It is still unclear whether or not it should be enabled by default for rocr iface. From your answer I extrapolate that it should be done at every `mr`?

Comment (iziemba) [reply include/ofi_hmem.h:123]: I guess I question if this is_dev_reg_enabled needed. I think the code in https://github.com/ofiwg/libfabric/blob/c37c44e5b485b5185de96c53453a7249ca3fd672/prov/rxm/src/rxm_domain.c#L602-L613 could be implemented as follows.

```
 	int ret = ofi_hmem_dev_register(rxm_mr->iface, attr->mr_iov->iov_base, attr->mr_iov->iov_len, (uint64_t *) &rxm_mr->hmem_handle);
	if (ret) {
		rxm_mr->hmem_flags = 0x0; 
 		rxm_mr->hmem_handle = NULL; 
	} else if (rxm_mr->iface == FI_HMEM_CUDA) {
		rxm_mr->hmem_flags = OFI_HMEM_DATA_GDRCOPY_HANDLE;
	}
```

Comment (thomasgillis) [reply include/ofi_hmem.h:123]: I don't have the history, but from the code one of the reason is that`ofi_hmem_dev_register` does not return an error if `FI_HMEM_CUDA_USE_GDRCOPY=0`.
@wenduwan @shijin-aws do you have more input?

Also, why only setting `OFI_HMEM_DATA_GDRCOPY_HANDLE` for cuda?
From Edgar's answer it would be useful for AMD as well.

Comment (iziemba) [reply include/ofi_hmem.h:123]: Making OFI_HMEM_DATA_GDRCOPY_HANDLE more generic for all HMEM seems to be the correct change.

Comment (wenduwan) [reply include/ofi_hmem.h:123]: @thomasgillis  Thanks for the heads up. I agree with @iziemba wrt if `ofi_hmem_dev_register` fails then the caller should have a fallback.

EFA only attempts `ofi_hmem_dev_register` if gdrcopy is enabled and available https://github.com/ofiwg/libfabric/blob/9f604d658277c3a4851506327cb7a75572622767/prov/efa/src/efa_mr.c#L253

Are you suggesting that we change `ofi_hmem_dev_register` so that it returns error if gdrcopy is disabled? I think that is also fine and backward compatible for EFA.

Comment (thomasgillis) [reply include/ofi_hmem.h:123]: I think making it fail would be more iface agnostic and help to comply to the if(error) don't do it strategy. I will try to make the changes asap then

Comment (thomasgillis) [reply include/ofi_hmem.h]: @wenduwan I have done the changes for the common part but only replaced the naming in EFA.
Is this the right way to do things to not break anything in EFA?

Comment (shijin-aws) [reply include/ofi_hmem.h]: Please make the efa change in the same commit where you renamed the macro. We shouldn't have a commit that cannot be compiled

File: include/ofi_mr.h [hunk 2]
```diff
@@ -270,10 +270,10 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
 
 /* libfabric internal flags starting from 60 to 63 */
 /**
- * OFI_HMEM_DATA_GDRCOPY_HANDLE indicates that hmem_data points to
- * a gdrcopy_handle data structure
+ * OFI_HMEM_DATA_DEV_REG_HANDLE indicates that hmem_data points to
+ * a dev reg data structure
```
[the remark this answers is not in the document]
Comment (thomasgillis) [reply include/ofi_mr.h:274]: is `OFI_HMEM_DATA_DEV_REG_HANDLE` the right name for this flag?

File: src/hmem_cuda.c [hunk 3]
```diff
@@ -314,7 +314,10 @@ int cuda_copy_from_dev(uint64_t device, void *dst, const void *src, size_t size)
 
 int cuda_dev_register(const void *addr, size_t size, uint64_t *handle)
 {
-	return cuda_gdrcopy_dev_register(addr, size, handle);
+	if (cuda_gdr_copy_enable()) {
+		return cuda_gdrcopy_dev_register(addr, size, handle);
+	}
+	return FI_EIO;
```
Comment (iziemba) [inline src/hmem_cuda.c:320]: This change be its own commit. Also, -FI_EIO should be returned.

File: include/ofi_mr.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (wenduwan) [reply include/ofi_mr.h]: It sounds reasonable to me.

**Comment by wenduwan** [comment]
bot:aws:retest

File: src/hmem.c [hunk 4]
```diff
@@ -404,17 +404,19 @@ static ssize_t ofi_copy_mr_iov(struct ofi_mr **mr, const struct iovec *iov,
 			hmem_data = NULL;
 		}
 
-		if (hmem_iface == FI_HMEM_CUDA && (hmem_flags & OFI_HMEM_DATA_GDRCOPY_HANDLE)) {
+		if (hmem_flags & OFI_HMEM_DATA_DEV_REG_HANDLE) {
 			/**
-			 * TODO: Fine tune the max data size to switch from gdrcopy to cudaMemcpy
-			 * Note: buf must be on the host since gdrcopy does not support D2D copy
+			 * TODO: Fine tune the max data size to switch from
+			 * def_reg copy to normal copy
```
Comment (wenduwan) [inline src/hmem.c:410]: I wasn't clear in my previous comment. I think we should remove this comment in src/hmem.c

This is the common code path and not concerned with tuning(which is up to the provider).

**Comment by wenduwan** [comment]
bot:aws:retest

**Review by wenduwan — APPROVED** [review]
Thank you!

**Review by shijin-aws — APPROVED** [review]

**Comment by shijin-aws** [comment]
need @aingerson's review for shm change

File: src/hmem_cuda.c [hunk 5]
```diff
@@ -314,7 +314,10 @@ int cuda_copy_from_dev(uint64_t device, void *dst, const void *src, size_t size)
 
 int cuda_dev_register(const void *addr, size_t size, uint64_t *handle)
 {
-	return cuda_gdrcopy_dev_register(addr, size, handle);
+	if (cuda_is_gdrcopy_enabled()) {
```
Comment (aingerson) [inline src/hmem_cuda.c:317]: drop brackets here

File: src/hmem_cuda.c [hunk 6]
```diff
@@ -314,7 +314,10 @@ int cuda_copy_from_dev(uint64_t device, void *dst, const void *src, size_t size)
...
+		return cuda_gdrcopy_dev_register(addr, size, handle);
+	}
+	return -FI_EIO;
```
Comment (aingerson) [inline src/hmem_cuda.c:320]: FI_ENOSYS makes more sense to me

**Review by aingerson — COMMENTED** [review]
A couple of nitpicky things. shm change looks fine. I would prefer the second patch to have a little bit in the commit message to provide context behind the change.
Thanks!

File: src/hmem_cuda.c (hunk 6 above)
Comment (thomasgillis) [reply src/hmem_cuda.c]: Hopefully the new commits address your change request. let me know if something else is needed at this point

**Comment by shijin-aws** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
@j-xiong is Intel CI failure related?

**Comment by j-xiong** [comment]
@shijin-aws There was a CI script change in one of the PRs merged yesterday which turned on some tests. Successful run of those tests depends on other fixes in that PR.  You could do a rebase to pick up those changes, or we can ignore the Intel CI failure.

**Comment by thomasgillis** [comment]
I can rebase if needed, let me know :-)

**Comment by j-xiong** [comment]
@thomasgillis Then rebase :)

**Review by aingerson — APPROVED** [review]
Thanks!

**Comment by thomasgillis** [comment]
@shijin-aws could you restart the aws ci? (hopefully for the last time)
@aingerson should i worry about the CI failure in `ze-shm-v3`?

**Comment by shijin-aws** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
@j-xiong can we add @thomasgillis to the organization? Then I don't need to manually trigger AWS CI

**Comment by aingerson** [comment]
@thomasgillis Node issue. All the other shm tests passed so it should be fine.

**Comment by j-xiong** [comment]
@shijin-aws Just invited @thomasgillis to the organization.

**Review by iziemba — APPROVED** [review]
